Description:
 When linking shared libraries, move -l options behind object files.
 Fixes build failures with -Wl,--as-needed for cases like "-lgnat foo.o".
Bug-Debian: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=705812
Author: Matthias Klose <doko@ubuntu.com>
Forwarded: yes

--- a/src/gprlib-build_shared_lib-nosymbols.adb
+++ b/src/gprlib-build_shared_lib-nosymbols.adb
@@ -145,11 +145,14 @@
       Add_Arg (Out_V);
       Lib_Index := Last_Arg;
 
-      --  The options
+      --  The options (anything except shared libraries).
 
-      for J in Options'Range loop
-         if Options (J) /= null and then Options (J).all /= "" then
-            Add_Arg (Options (J));
+      for Opt of Options loop
+         if Opt /= null and then Opt.all /= ""
+           and then (Opt'Length < 2
+                       or else Opt (Opt'First .. Opt'First + 1) /= "-l")
+         then
+            Add_Arg (Opt);
          end if;
       end loop;
 
@@ -293,6 +296,16 @@
          Add_Arg (Library_Options_Table.Table (J));
       end loop;
 
+      --  The options (shared libraries)
+
+      for Opt of Options loop
+         if Opt /= null and then Opt'Length > 2
+           and then Opt (Opt'First .. Opt'First + 1) = "-l"
+         then
+            Add_Arg (Opt);
+         end if;
+      end loop;
+
       Display_Linking_Command;
 
       --  Check if a response file is needed
